Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift

```diff
@@ -37,88 +37,444 @@ extension PathHierarchy {
-                let kindWithoutLanguage = String(kind.dropFirst(languagePrefix.count))
+                let kindWithoutLanguage = kind.dropFirst(languagePrefix.count)
                 return PathComponent(full: full, name: name, kind: kindWithoutLanguage, hash: hash)
             }
         }
         return PathComponent(full: full, name: name, kind: nil, hash: hash)
     }
     
-    static func joined<PathComponents>(_ pathComponents: PathComponents) -> String where PathComponents: Sequence, PathComponents.Element == PathComponent {
-        return pathComponents.map(\.full).joined(separator: "/")
+    static func split(_ path: String) -> (componentSubstrings: [Substring], isAbsolute: Bool) {
+        var components: [Substring] = self.split(path)
+        
+        let isAbsolute: Bool
+        if path.first == PathComponentScanner.separator {
+            guard let maybeModuleName = components.first?.dropFirst(), !maybeModuleName.isEmpty else {
+                return ([], true)
+            }
+            isAbsolute = maybeModuleName.isValidModuleName()
+            assert(NodeURLGenerator.Path.documentationFolderName.isValidModuleName())
+            assert(NodeURLGenerator.Path.tutorialsFolderName.isValidModuleName())
+            if isAbsolute {
+                _ = components[components.startIndex].removeFirst()
+            }
+        } else {
+            let name = components.first.map(String.init)
+            isAbsolute = name == NodeURLGenerator.Path.documentationFolderName
+                      || name == NodeURLGenerator.Path.tutorialsFolderName
+        }
+        
+        return (components, isAbsolute)
+    }
+    
+    private static func split(_ path: String) -> [Substring] {
+        var result = [Substring]()
+        var scanner = PathComponentScanner(path[...])
+        
+        let anchorResult = scanner.scanAnchorComponent()
+        
+        while !scanner.isEmpty {
+            let component = scanner.scanPathComponent()
+            if !component.isEmpty {
+                result.append(component)
+            }
+        }
+        
+        if let anchorResult = anchorResult{
+            result.append(anchorResult)
+        }
+
+        return result
+    }
+}
+
+private struct PathComponentScanner {
+    private var remaining: Substring
+    
+    static let separator: Character = "/"
+    private static let anchorSeparator: Character = "#"
+    private static let operatorEnd: Character = ")"
+    
+    init(_ original: Substring) {
+        remaining = original
+    }
+    
+    var isEmpty: Bool {
+        remaining.isEmpty
+    }
+    
+    mutating func scanPathComponent() -> Substring {
+        // If the next component is an operator, parse the full operator before splitting on "/" ("/" may appear in the operator name)
+        if remaining.unicodeScalars.prefix(3).allSatisfy(\.isValidOperatorHead) {
```

Comment:
It's meant to be the shortest possible[^1] operator name[^2]; one operator head character followed by `()` for no arguments.

We could shorten it to just two characters because we only need to find if the next component isn't a module with a "/" prefix. 

[^1]: Having no arguments doesn't really make sense for a real operator since there's no value to operate on. Even `!` has one argument.
[^2]: It's really too permissive for real operator names since the operator head supports more characters than the rest of the operator name does.

Reply:
Maybe add a comment to this effect (someday... not urgent) - or define a let constant with a descriptive name, so the number is less magical.

Reply:
I'll probably revisit this piece of code for rdar://119616317